File: roles/prover.py
import random
from typing import List, Tuple, Dict

from py_ecc.bls12_381 import curve_order

from common.datastructures import DPDPTags, DPDPProof, Block
from Crypto.dpdp import dPDP
from utils import log_msg

class Prover:
    """
    负责为存储的数据生成dPDP证明。
    """
    def __init__(self, node_id: str):
        self.node_id = node_id

    def prove(
        self,
        file_id: str,
        indices: List[int],
        file_chunks: Dict[int, bytes],
        file_tags: DPDPTags,
        block: Block
    ) -> Tuple[DPDPProof, List[Tuple[int, int]], List[Tuple[int, int, tuple]]]:
        """
        为指定轮次(由区块参数确定)生成dPDP聚合证明与未聚合贡献。

        :param block: 区块信息，提供(prev_hash, timestamp)以派生公开挑战
        :param file_id: 文件的ID。
        :param indices: 兼容参数（忽略），挑战由区块确定。
        :param file_chunks: 文件中所有分片的原始数据。
        :param file_tags: 文件中所有分片的dPDP标签。
        :return: (聚合证明, 挑战[(i,v_i)], 贡献[(i, mu_i, sigma_i)])
        """
        # 1. 生成公开挑战集合
        challenge = dPDP.gen_chal(block.prev_hash, block.timestamp, file_tags)

        # 2. 准备数据子集
        challenged_chunks = {i: file_chunks[i] for i, _ in challenge if i in file_chunks}

        # 3. 生成未聚合贡献
        contributions = dPDP.gen_contributions(file_tags, challenged_chunks, challenge)

        # 4. 基于挑战/标签生成聚合证明
        proof = dPDP.gen_proof(file_tags, challenged_chunks, challenge)
        log_msg("DEBUG", "dPDP", self.node_id, f"为文件 {file_id} 生成了dPDP证明与未聚合贡献")

        # 5. 返回三元组
        # 注意：sigma_i为G1点元组，已可序列化（由调用方处理）
        return proof, challenge, contributions

File: src/roles/prover.rs
// 导入标准库的 HashMap
use std::collections::HashMap;

// 导入大整数库
use num_bigint::BigUint;

// 导入项目内的数据结构和密码学模块
use crate::common::datastructures::{DPDPProof, DPDPTags};
use crate::crypto::dpdp::DPDP;
use crate::utils::log_msg;

/// dPDP 挑战向量类型别名，元素为 (块索引, 挑战系数)。
pub type ChallengeVector = Vec<(usize, BigUint)>;

/// 每个挑战对应的贡献值，包含块索引、加权系数和缩放后的标签。
pub type ContributionVector = Vec<(usize, BigUint, Vec<u8>)>;

/// Prover 结构体定义了证明者的角色
/// 证明者的核心职责是响应 dPDP（动态可证明数据拥有权）挑战，
/// 生成一个密码学证明，以证实自己确实存储了特定的文件数据。
pub struct Prover {
    // 证明者所属节点的 ID
    pub node_id: String,
}

impl Prover {
    /// 创建一个新的 Prover 实例
    pub fn new(node_id: String) -> Self {
        Self { node_id }
    }

    /// 为给定的文件生成 dPDP 证明
    ///
    /// # Arguments
    /// * `file_id` - 正在为其生成证明的文件的 ID
    /// * `file_chunks` - 文件的实际数据块内容
    /// * `file_tags` - 文件的 dPDP 标签，这是预先计算好的用于生成证明的元数据
    /// * `prev_hash` - 上一个区块的哈希，用作生成挑战的盐，确保挑战的不可预测性
    /// * `timestamp` - 时间戳，同样用于生成挑战
    /// * `challenge_size` - 可选参数，指定挑战中要抽样的块的数量
    ///
    /// # Returns
    /// 一个元组，包含：
    /// * `DPDPProof` - 生成的聚合 dPDP 证明
    /// * `Vec<(usize, BigUint)>` - 生成的挑战，格式为 (索引, 随机值) 的列表
    /// * `Vec<(usize, BigUint, Vec<u8>)>` - 未聚合的贡献值，用于更新文件的状态（例如时间戳或版本）
    pub fn prove(
        &self,
        file_id: &str,
        file_chunks: &HashMap<usize, Vec<u8>>,
        file_tags: &DPDPTags,
        prev_hash: &str,
        timestamp: u64,
        challenge_size: Option<usize>,
    ) -> (DPDPProof, ChallengeVector, ContributionVector) {
        // 1. 根据上下文（前一个块哈希、时间戳）和文件标签生成一个确定性的随机挑战
        let challenge = DPDP::gen_chal(prev_hash, timestamp, file_tags, challenge_size);

        // 2. 使用文件块、标签和挑战来生成聚合的 dPDP 证明
        let proof = DPDP::gen_proof(file_tags, file_chunks, &challenge);

        // 3. 生成用于更新文件状态的“贡献值”
        // 这些值是证明过程的副产品，但对于维护文件的版本和状态至关重要
        let contributions = DPDP::gen_contributions(file_tags, file_chunks, &challenge);

        log_msg(
            "DEBUG",
            "dPDP",
            Some(self.node_id.clone()),
            &format!("为文件 {} 生成了dPDP证明与未聚合贡献", file_id),
        );

        // 返回证明、挑战和贡献值
        (proof, challenge, contributions)
    }
}
